implementations/rust/ockam/ockam_app_lib/src/incoming_services/state.rs

```diff
@@ -0,0 +1,365 @@
+            project.name = project.id.clone();
+        }
+        Ok(enrollment_ticket)
+    }
+
+    pub fn relay_name(&self) -> String {
+        let bare_relay_name = self.access_details.shared_node_identity.to_string();
+        format!("forward_to_{bare_relay_name}")
+    }
+
+    /// Returns the name of the inlet within the node, for now it's a constant
+    pub fn inlet_name(&self) -> &str {
+        "app-inlet"
+    }
+
+    pub fn service_route(&self) -> ockam::Result<String> {
+        if let Some(project) = self.enrollment_ticket()?.project.as_ref() {
+            let project_id = &project.id;
+            let relay_name = self.relay_name();
+            let service_name = match self.original_name() {
+                Ok(name) => name,
+                Err(_) => {
+                    warn!("Failed to get service name from access details");
+                    return Err(ApiError::core(
+                        "Failed to get service name from access details",
+                    ));
+                }
+            };
+            Ok(format!(
+                "/project/{project_id}/service/{relay_name}/secure/api/service/{service_name}"
+            ))
+        } else {
+            Err(ApiError::core("No project id found in enrollment ticket"))
+        }
+    }
+
+    pub fn local_node_name(&self) -> ockam::Result<String> {
+        if let Some(project) = self.enrollment_ticket()?.project.as_ref() {
+            let project_id = &project.id;
+            let id = &self.id;
+            Ok(format!("ockam_app_{project_id}_{id}"))
+        } else {
+            Err(ApiError::core("No project id found in enrollment ticket"))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::incoming_services::PersistentIncomingServiceState;
+    use crate::state::AppState;
+    use ockam::identity::{Identifier, OneTimeCode};
+    use ockam_api::cli_state::CliState;
+    use ockam_api::cloud::share::{
+        InvitationWithAccess, ReceivedInvitation, RoleInShare, ServiceAccessDetails, ShareScope,
+    };
+    use ockam_api::config::lookup::ProjectLookup;
+    use ockam_api::identity::EnrollmentTicket;
+    use std::str::FromStr;
+
+    #[test]
+    fn test_inlet_data_from_invitation() {
+        let mut invitation = InvitationWithAccess {
+            invitation: ReceivedInvitation {
+                id: "invitation_id".to_string(),
+                expires_at: "2020-09-12T15:07:14.00".to_string(),
+                grant_role: RoleInShare::Admin,
+                owner_email: "owner_email".to_string(),
+                scope: ShareScope::Project,
+                target_id: "target_id".to_string(),
+                ignored: false,
+            },
+            service_access_details: None,
+        };
+
+        // Skipping an invitation without service access details
+        let app_state = AppState::new(None, None, Some(CliState::test().unwrap())).unwrap();
+
+        app_state.context().runtime().block_on(async move {
```

Comment:
Can we run this test as a `tokio::test` instead of blocking here?

Reply:
no, creating the context of the ockam app would create another tokio runtime resulting in a panic

Reply:
I've created a branch with a commit showing how we can use the `ockam::test` macro to have a proper `Context` being provided to run the test if you want to have a look at it: `davide-baldo/app-persistent-service-state-with-ockam-test`

Reply:
Sorry @davide-baldo I must have forgotten to push. Please try again.